src/Internal/Payment/Storage.php

```diff
@@ -0,0 +1,175 @@
+	 * @param PaymentMethodService $payment_method_service PM service.
+	 */
+	public function __construct(
+		StateFactory $state_factory,
+		PaymentMethodService $payment_method_service
+	) {
+		$this->state_factory          = $state_factory;
+		$this->payment_method_service = $payment_method_service;
+	}
+
+	/**
+	 * Enables payment storage.
+	 *
+	 * IMPORTANT: By default, this class does not touch the database.
+	 * Once enabled, it will, but this should only be used for tests
+	 * before the final storage method is determined.
+	 */
+	public function enable() {
+		$this->enabled = true;
+	}
+
+	/**
+	 * Loads or creates the payment object for an order,
+	 * returning the one true payment object.
+	 *
+	 * @param WC_Order $order Order that requires/has payment.
+	 * @return State          Current state of the order payment.
+	 */
+	public function get_order_payment( WC_Order $order ) {
+		$payment     = new Payment( $order );
+		$state_class = InitialState::class;
+
+		if ( $this->enabled ) {
+			$data = $order->get_meta( self::ORDER_META_KEY );
+			if ( is_array( $data ) && ! empty( $data ) ) {
+				$state_class = $data['state'];
+				$this->import_payment_data( $payment, $data );
+			}
+		}
+
+		$state = $this->state_factory->create_state( $state_class );
+		$state->set_context( $payment );
+
+		return $state;
+	}
+
+	/**
+	 * Prepares a payment for storage.
+	 *
+	 * @param State $payment_state The current payment state.
+	 */
+	public function save_order_payment( State $payment_state ): void {
+		if ( ! $this->enabled ) {
+			return;
+		}
+
+		$payment = $payment_state->get_context();
+		$data    = $this->extract_payment_data( $payment );
+
+		// State is not a part of the payment object, but should be saved.
+		$data['state'] = get_class( $payment_state );
+
+		// Simply store as meta, and save the order immediately.
+		$order = $payment->get_order();
+		$order->update_meta_data( self::ORDER_META_KEY, $data );
+		$order->save();
+	}
+
+	/**
+	 * Delete all stored payment data from an order.
+	 *
+	 * WARNING: This method is destructive!
+	 * It is useful for development, but should be used with caution.
+	 *
+	 * @param WC_Order $order Order that might have payment.
+	 */
+	public function cleanup_order_payment( WC_Order $order ) {
+		$order->delete_meta_data( self::ORDER_META_KEY );
+		$order->save();
+	}
+
+	/**
+	 * Extracts all data from a payment object in the form of simple values.
+	 *
+	 * @param Payment $payment Payment object.
+	 * @return array
+	 */
+	private function extract_payment_data( Payment $payment ): array {
+		$data = [];
+
+		$payment_method         = $payment->get_payment_method();
+		$data['payment_method'] = $payment_method ? $payment_method->get_data() : null;
+
+		$data['flags'] = [];
+		foreach ( Flag::get_all() as $flag ) {
+			if ( $payment->has_flag( $flag ) ) {
+				$data['flags'][] = $flag->__toString();
+			}
+		}
+
+		return $data;
+	}
```

Comment:
Serializing and deserializing payment data might make sense outside the payment object, keeping it as lightweight as possible.